File: javascript/lib/gherkin/ast_builder.js
var AstNode = require('./ast_node');

module.exports = function AstBuilder () {

  var stack = [new AstNode('None')];

  this.startRule = function (ruleType) {
    stack.push(new AstNode(ruleType));
  };

  this.endRule = function (ruleType) {
    var node = stack.pop();
    var transformedNode = transformNode(node);
    currentNode().add(node.ruleType, transformedNode);
  };

  this.build = function (token) {
    currentNode().add(token.matchedType, token);
  };

  this.getResult = function () {
    return currentNode().getSingle('Feature');
  };

  function currentNode () {
    return stack[stack.length - 1];
  }

  function transformNode(node) {
    return node;
  }

};

File: ruby/lib/gherkin/ast_builder.rb
require 'gherkin/ast_node'

module Gherkin
  class AstBuilder
    attr_reader :stack

    def initialize
      @stack = [AstNode.new('None')]
    end

    def start_rule(rule_type)
      stack.push AstNode.new(rule_type)
    end

    def end_rule(rule_type)
      node = stack.pop
      current_node.add(node.rule_type, transform_node(node))
    end

    def build(token)
      current_node.add(token.matched_type, token)
    end

    def current_node
      stack.last
    end
  end
end
